Sources/MessagingPush/PushHandling/AutomaticPushClickHandling.swift

```diff
@@ -0,0 +1,51 @@
+ 1. Getting an instance of `UNUserNotificationCenterDelegate` from `UNUserNotificationCenter.current().delegate`.
+ 2. Calling `userNotificationCenter(_:didReceive:withCompletionHandler:)` on the delegate.
+
+ This is simple when the CIO SDK is the only SDK in an app that needs to get notified when a push is clicked on for a host app. When a customer has installed multiple SDKs into their app that all want to handle when a push is clicked, that's when this gets complex. That's because iOS has a restriction that only 1 object can be set as the `UNUserNotificationCenter.current().delegate`. The CIO SDK when it's initialized can set itself as this delegate instance, but then another SDK can set itself as the delegate instead so the CIO SDK is then no longer called when a push is clicked.
+
+ To solve this problem, the CIO SDK performs this logic:
+ 1. The CIO SDK forces itself to always be the only `UNUserNotificationCenter.current().delegate` instance of the host iOS app.
+
+ The SDK does this via swizzling. When `UNUserNotificationCenter.current().delegate` setter gets called, the CIO SDK gets notified that this event happened. When a new delegate gets set, the CIO SDK reverses this action by resetting the CIO SDK `UNUserNotificationCenterDelegate` instance as the delegate.
+
+ Related code for this logic:
+ * `iOSPushEventListener` - where swizzling is.
+ * `iOSPushEventListener` - is the CIO SDK instance of `UNUserNotificationCenterDelegate`.
+
+ 2. When the CIO SDK's instance of `UNUserNotificationCenterDelegate` is called, the SDK forwards that push click event to all other `UNUserNotificationCenterDelegate` instances registered to the app. This allows other SDKs the ability to handle a push click event for pushes not sent by CIO.
+
+ Related code for this logic:
+ * `NotificationCenterDelegateProxy` - stores all other `UNUserNotificationCenterDelegate` instances that have been registered with the host app.
+ * `iOSPushEventListener` - is what calls the proxy class when a push is clicked that was not sent by CIO.
+
+ The goal of this feature is:
+ 1. The customer does not need to interact with `UNUserNotificationCenter` themselves to get the CIO SDK to process when a push is clicked. The CIO SDK should be able to set this up itself.
+ 2. The CIO SDK should be able to stay compatible with other SDKs that also want to handle push click events. A customer should be able to install 2+ push notification SDKs in an app and all of them are able to work, even though iOS only allows 1 `UNUserNotificationCenterDelegate` instnace to be set in the app.
+ */
+@available(iOSApplicationExtension, unavailable)
+protocol AutomaticPushClickHandling: AutoMockable {
+    func start()
+}
+
+@available(iOSApplicationExtension, unavailable)
+// sourcery: InjectRegister = "AutomaticPushClickHandling"
+class AutomaticPushClickHandlingImpl: AutomaticPushClickHandling {
```

Comment:
Appreciate all the notes you added, really helpful 👏🏻 But I'm not sure why we introduced `AutomaticPushClickHandling` as it looks like `PushEventListener` could alone handle this?

Reply:
> But I'm not sure why we introduced AutomaticPushClickHandling as it looks like PushEventListener could alone handle this?

That is a fair question and one that I was also asking myself when I created the file. 

The decision behind this file is to house the comments that I wrote describing this feature. I felt the `PushEventListener` file was very large already and I didn't want to make it any larger in size by putting the documentation in there. 

I liked the idea of having a file named "AutomaticPushClickHandling" and the documentation of that feature going inside of this file. That seems easy to find for future developers working in the code base. Felt easier to find and maintain compared to adding docs somewhere else like a notion doc. 

If we don't like this idea, perhaps creating a `README.md` inside of the `Sources/MessagingPush/PushHandling/` or `Sources/MessagingPush/` directory would be more welcome?

Reply:
All makes sense to me. I would have preferred the `README.md` inside `Sources/MessagingPush/`, but I'm okay with this too. And since you have already put efforts writing and testing it, I think we should stick with it for now.